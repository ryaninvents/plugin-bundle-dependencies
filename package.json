{
  "name": "@ryaninvents/plugin-bundle-dependencies",
  "version": "0.0.0-semantically-released",
  "description": "Plugin for @pika/pack to zip dependencies for Node.js",
  "author": "Ryan Kennedy <ryaninvents@users.noreply.github.com>",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "@pika/types": "^0.9.2",
    "del": "^5.1.0",
    "execa": "^1.0.0",
    "fs-extra": "^7.0.1",
    "gulp": "^4.0.0",
    "gulp-rename": "^1.4.0",
    "gulp-zip": "^4.2.0",
    "has-yarn": "^1.0.0",
    "temporarily": "^1.0.5"
  },
  "@pika/pack": {
    "pipeline": [
      [
        "@pika/plugin-standard-pkg",
        {
          "exclude": [
            "__tests__/**"
          ]
        }
      ],
      [
        "@pika/plugin-build-node",
        {
          "minNodeVersion": "8"
        }
      ]
    ]
  },
  "devDependencies": {
    "@babel/preset-env": "^7.3.1",
    "@pika/pack": "^0.5.0",
    "@pika/plugin-build-node": "^0.9.2",
    "@pika/plugin-standard-pkg": "^0.9.2",
    "@semantic-release/changelog": "^3.0.2",
    "@semantic-release/commit-analyzer": "^6.1.0",
    "@semantic-release/git": "^7.0.8",
    "@semantic-release/github": "^5.2.10",
    "@semantic-release/npm": "^5.1.4",
    "@semantic-release/release-notes-generator": "^7.1.4",
    "@types/gulp": "^4.0.5",
    "@types/gulp-zip": "^4.0.0",
    "@types/node": "^13.9.8",
    "eslint": "^5.13.0",
    "eslint-config-semistandard": "^13.0.0",
    "eslint-config-standard": "^12.0.0",
    "eslint-plugin-import": "^2.16.0",
    "eslint-plugin-node": "^8.0.1",
    "eslint-plugin-promise": "^4.0.1",
    "eslint-plugin-standard": "^4.0.0",
    "jest": "^25.2.4",
    "ls-archive": "^1.3.1",
    "prettier": "^2.0.2",
    "semantic-release": "^17.2.3",
    "typescript": "^3.3.3"
  },
  "eslintConfig": {
    "extends": "semistandard",
    "overrides": [
      {
        "files": [
          "**/__tests__/**",
          "*.test.js"
        ],
        "env": {
          "jest": true
        }
      }
    ],
    "rules": {
      "space-before-function-paren": "off"
    }
  },
  "jest": {
    "roots": [
      "<rootDir>/src"
    ]
  },
  "babel": {
    "presets": [
      [
        "@babel/preset-env",
        {
          "targets": {
            "node": 8
          },
          "modules": false
        }
      ]
    ],
    "env": {
      "test": {
        "presets": [
          [
            "@babel/preset-env",
            {
              "targets": {
                "node": 8
              }
            }
          ]
        ]
      }
    }
  },
  "release": {
    "branch": "develop",
    "plugins": [
      "@semantic-release/commit-analyzer",
      "@semantic-release/release-notes-generator",
      "@semantic-release/changelog",
      [
        "@semantic-release/npm",
        {
          "pkgRoot": "./pkg"
        }
      ],
      "@semantic-release/github",
      [
        "@semantic-release/git",
        {
          "assets": [
            "CHANGELOG.md"
          ],
          "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
        }
      ]
    ]
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "jest",
    "build": "pika build",
    "semantic-release": "semantic-release"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/ryaninvents/plugin-bundle-dependencies.git"
  }
}
