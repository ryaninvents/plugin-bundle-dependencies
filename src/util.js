import hasYarn from 'has-yarn';
import execa from 'execa';
import { resolve } from 'path';
import fs from 'fs-extra';

const LOCAL_PACKAGE_PREFIX = 'file:';

export async function npmPack (pkgPath, workingDir) {
  const { stdout } = await execa('npm', ['pack', '--json', pkgPath], {
    cwd: workingDir
  });
  const [output] = JSON.parse(stdout);
  return resolve(workingDir, output.filename);
}

export function getPackageManagerCommand ({ cwd, options }) {
  if (options.packageManager) {
    const { packageManager } = options;
    if (packageManager === 'npm') {
      return { bin: 'npm', args: ['install'] };
    } else if (packageManager === 'yarn') {
      return { bin: 'yarn', args: [] };
    }
    const errors = [];
    const { bin } = packageManager;
    if (typeof bin !== 'string') {
      errors.push(`Expected \`packageManager.bin\` to be a string (got ${typeof bin})`);
    }
    const { args = [] } = packageManager;
    if (args) {
      if (!Array.isArray(args) || args.some((arg) => typeof arg !== 'string')) {
        errors.push(`Expected \`packageManager.args\` to be an array of arguments`);
      }
    }
    return { bin, args };
  }
  if (hasYarn(cwd)) {
    return { bin: 'yarn', args: [] };
  }
  return { bin: 'npm', args: ['install'] };
}

export async function createManifest (tempdir, { cwd, manifest }) {
  const updatedManifest = { ...manifest };
  const originalDeps = manifest.dependencies || {};
  updatedManifest.dependencies = { ...originalDeps };

  await Promise.all(Object.keys(originalDeps).map(async (pkgName) => {
    const pkgVersion = originalDeps[pkgName];
    if (!pkgVersion.startsWith(LOCAL_PACKAGE_PREFIX)) return;
    const absolutePath = resolve(cwd, pkgVersion.slice(LOCAL_PACKAGE_PREFIX.length));
    const bundledAbsolutePath = await npmPack(absolutePath, cwd);
    updatedManifest.dependencies[pkgName] = `${LOCAL_PACKAGE_PREFIX}${bundledAbsolutePath}`;
  }));

  await fs.writeFile(resolve(tempdir, 'package.json'), JSON.stringify(updatedManifest, null, 2));
}

export async function installDependencies (tempdir, { cwd, options = {} }) {
  const { bin, args } = getPackageManagerCommand({ cwd, options });
  const env = {
    NODE_ENV: 'production',
    ...(options.env || {})
  };
  try {
    await execa(bin, args, {
      cwd: tempdir,
      env,
      stdio: options.stdio || 'ignore'
    });
  } catch (err) {
    err.message += `\n${err.stdout}`;
    throw err;
  }
}
