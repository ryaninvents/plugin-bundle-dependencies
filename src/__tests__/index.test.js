import fs from 'fs-extra';
import { join, relative, resolve } from 'path';
import { dir as createTempDir, file as createTempFile } from 'temporarily';
import del from 'del';
import execa from 'execa';
import archive from 'ls-archive';
import { installDependencies } from '../util';
import { name as thisPackageName } from '../../package.json';

const DEFAULT_PKG_JSON = {
  name: 'test-pkg',
  private: true,
  version: '0.0.0-develop',
  scripts: {
    build: 'pack build'
  },
  '@pika/pack': {
    'pipeline': [
      [
        '@pika/plugin-standard-pkg'
      ],
      [
        '@pika/plugin-build-node',
        {
          'minNodeVersion': '8'
        }
      ],
      [
        thisPackageName
      ]
    ]
  },
  'devDependencies': {
    '@pika/pack': '^0.3.1',
    '@pika/plugin-build-node': '^0.3.10',
    '@pika/plugin-standard-pkg': '^0.3.10',
    [thisPackageName]: resolve(__dirname, '../../pkg')
  }
};

describe('@ryaninvents/plugin-bundle-dependencies', () => {
  let createdDirs = [];

  async function initRepo ({ packageJson: pkg = DEFAULT_PKG_JSON, packageManager = 'npm', stdio } = {}) {
    const { filepath: workingDir } = createTempDir(
      { name: 'test-repo-{wwwwdddd}' },
      [createTempDir({ name: 'src' }, [
        createTempFile({ name: 'index.js', data: '// Hello world' })
      ])]
    );
    let packageJson = pkg;
    if (typeof packageJson === 'function') {
      packageJson = packageJson(workingDir);
    }
    await fs.writeFile(join(workingDir, 'package.json'), JSON.stringify(packageJson, null, 2));
    await installDependencies(workingDir, {
      cwd: workingDir,
      options: { env: { NODE_ENV: 'development' } },
      packageManager
    });
    createdDirs.push(workingDir);
    return { workingDir };
  }

  afterEach(async () => {
    await del(createdDirs, { force: true });
    createdDirs = [];
  });

  it('should create a repo and install node_modules', async () => {
    const { workingDir } = await initRepo({
      packageJson: {
        ...DEFAULT_PKG_JSON,
        dependencies: {
          del: 'latest'
        }
      },
      stdio: 'inherit'
    });
    expect(async () => fs.access(`${workingDir}/package.json`, fs.constants.F_OK))
      .not.toThrow();
    expect(async () => fs.access(`${workingDir}/src`, fs.constants.X_OK))
      .not.toThrow();
    expect(async () => fs.access(`${workingDir}/node_modules`, fs.constants.X_OK))
      .not.toThrow();
  }, 60e3);

  it('should correctly resolve local dependencies', async () => {
    const libProj = await initRepo({
      packageJson: {
        ...DEFAULT_PKG_JSON,
        name: 'test-library'
      },
      stdio: 'inherit'
    });
    await fs.writeFile(`${libProj.workingDir}/index.js`, '// Hello world');
    const mainProj = await initRepo({
      packageJson: (mainDir) => ({
        ...DEFAULT_PKG_JSON,
        dependencies: {
          'test-library': `file:${relative(mainDir, libProj.workingDir)}`
        }
      }),
      stdio: 'inherit'
    });
    expect(async () => fs.access(`${mainProj.workingDir}/node_modules/test-library`, fs.constants.X_OK))
      .not.toThrow();
    const contents = String(await fs.readFile(`${mainProj.workingDir}/node_modules/test-library/index.js`));
    expect(contents).toBe('// Hello world');
  }, 60e3);

  it('should create a zip file with node_modules', async () => {
    const { workingDir } = await initRepo({
      packageJson: {
        ...DEFAULT_PKG_JSON,
        dependencies: {
          // Use a couple of packages known to basically be one-liners
          // in order to test quickly
          'is-sorted': 'latest',
          'map-obj': 'latest'
        }
      },
      stdio: 'inherit'
    });
    await execa('npm', ['run', 'build'], {
      env: { NODE_ENV: 'production' },
      cwd: workingDir
    });
    expect(async () => fs.access(`${workingDir}/pkg/package.json`, fs.constants.F_OK))
      .not.toThrow();
    expect(async () => fs.access(`${workingDir}/pkg/dist-dependencies.zip`, fs.constants.F_OK))
      .not.toThrow();
    const archiveEntries = await new Promise((resolve, reject) => {
      archive.list(`${workingDir}/pkg/dist-dependencies.zip`, (err, results) => {
        if (err) return reject(err);
        return resolve(results.map(file => file.getPath()).sort());
      });
    });
    expect(archiveEntries).toMatchSnapshot();
  }, 60e3);

  it('should observe `prefix` option', async () => {
    const { workingDir } = await initRepo({
      packageJson: {
        ...DEFAULT_PKG_JSON,
        dependencies: {
          // Use a couple of packages known to basically be one-liners
          // in order to test quickly
          'is-sorted': 'latest',
          'map-obj': 'latest'
        },
        '@pika/pack': {
          'pipeline': [
            [
              '@pika/plugin-standard-pkg'
            ],
            [
              '@pika/plugin-build-node',
              {
                'minNodeVersion': '8'
              }
            ],
            [
              thisPackageName,
              { prefix: 'nodejs/node_modules' }
            ]
          ]
        }
      },
      stdio: 'inherit'
    });
    await execa('npm', ['run', 'build'], {
      env: { NODE_ENV: 'production' },
      cwd: workingDir
    });
    expect(async () => fs.access(`${workingDir}/pkg/package.json`, fs.constants.F_OK))
      .not.toThrow();
    expect(async () => fs.access(`${workingDir}/pkg/dist-dependencies.zip`, fs.constants.F_OK))
      .not.toThrow();
    const archiveEntries = await new Promise((resolve, reject) => {
      archive.list(`${workingDir}/pkg/dist-dependencies.zip`, (err, results) => {
        if (err) return reject(err);
        return resolve(results.map(file => file.getPath()).sort());
      });
    });
    expect(archiveEntries).toMatchSnapshot();
  }, 60e3);
}, 60e3);
