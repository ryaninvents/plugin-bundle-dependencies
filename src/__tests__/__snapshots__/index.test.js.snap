// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ryaninvents/plugin-bundle-dependencies should correctly resolve local dependencies 1`] = `
Array [
  "test-library",
  "test-library/index.js",
  "test-library/package.json",
  "test-library/src",
  "test-library/src/index.js",
]
`;

exports[`@ryaninvents/plugin-bundle-dependencies should create a zip file with node_modules 1`] = `
Array [
  "is-sorted",
  "is-sorted/.travis.yml",
  "is-sorted/LICENSE",
  "is-sorted/README.md",
  "is-sorted/index.js",
  "is-sorted/package.json",
  "is-sorted/test",
  "is-sorted/test/fixtures.json",
  "is-sorted/test/index.js",
  "map-obj",
  "map-obj/index.js",
  "map-obj/license",
  "map-obj/package.json",
  "map-obj/readme.md",
]
`;

exports[`@ryaninvents/plugin-bundle-dependencies should observe \`prefix\` option 1`] = `
Array [
  "nodejs/node_modules",
  "nodejs/node_modules/is-sorted",
  "nodejs/node_modules/is-sorted/.travis.yml",
  "nodejs/node_modules/is-sorted/LICENSE",
  "nodejs/node_modules/is-sorted/README.md",
  "nodejs/node_modules/is-sorted/index.js",
  "nodejs/node_modules/is-sorted/package.json",
  "nodejs/node_modules/is-sorted/test",
  "nodejs/node_modules/is-sorted/test/fixtures.json",
  "nodejs/node_modules/is-sorted/test/index.js",
  "nodejs/node_modules/map-obj",
  "nodejs/node_modules/map-obj/index.js",
  "nodejs/node_modules/map-obj/license",
  "nodejs/node_modules/map-obj/package.json",
  "nodejs/node_modules/map-obj/readme.md",
]
`;
